### PR TITLE
Restricts WIP pull requests and adds some guidelines regarding code unfinished in other ways

### PR DESCRIPTION
**Pull request by ninjanomnom**
We've had many bad code changes get through pull requests because of the following:

1. Large and complex pull request gets created
2. It is flawed and is a massive pain to review to deal with those flaws
3. Reviews are nonetheless attempted (or not) and some things improve but is overall still flawed
4. After a while it gets pity merged or merged because we hope it will be improved later and because the maintainers reviewing it have become exhausted by that pr
5. We now have another pile of shitcode that never gets improved

Pull requests in the past tried by people who don't have enough experience with writing clean code depend on maintainers pointing out every flaw. This encourages contributors to not really care about improving themselves and just pr things that work technically. If they want to improve the very concept of how they achieve these large changes they should discuss these things ahead of time with others.

Merging these large changes that barely work hurts us more than we lose by not allowing these changes on principle. In the end we may even end up with our contributors caring more about the quality of their code.

### CODE REVIEW

**Comment by optimumtact** [comment]
This is already true but I have never seen you close a single PR, so how would this help?

It frustrates me that you complain about this but although you review things you just let them sit on the tracker.

Im not a mind reader ninja, and I can't tell when you think code is overrall not worth having unless you're willing to say so.

**Comment by ninjanomnom** [comment]
I close prs often enough but I tend to try to stay out of gameplay decisions outside of what I'm intimate with. The WIP pr rule is certainly new. Also @optimumtact it's not like I'm a mind reader either, I certainly didn't know that closing prs that are conceptually fine but massively flawed in code is something we approved of doing since it isn't in the contribution guidelines.

**Comment by kevinz000** [comment]
Oh hey look it's me

**Comment by optimumtact** [comment]
That's like the first thing maintainers are supposed to be doing

**Comment by ninjanomnom** [comment]
Where was that in orientation, I just uphold the contribution guidelines and my own sense of code quality

**Comment by optimumtact** [comment]
We don't have an orientation but I don't see how it isn't obvious that if code is not worth maintaining it shouldn't be merged

**Comment by MrStonedOne** [comment]
To be fair, there is a very valid point hidden in that in that only the first time a big pr is reviewed by a maintainer is it *really* reviewed. in latter re-reviews, code-reread-fatigue sets in.

So this is still something that would be worthy of trying to encourage away.

Maybe start with just a more general/passive `While we have no issue helping contributors (and especially new contributors) bring small contributions up to standards via the pull request review process, Larger contributions are expected to pass a higher bar of completeness and code quality *before* opening a pull request. This helps us focus our attention more efficiently`?

**Comment by lordpidey** [comment]
How will this interact with draft PRs?

**Comment by MrPerson** [comment]
You shouldn't complain that people don't know something if you never tell it to them, no matter how "obvious" it might seem to you.

**Comment by optimumtact** [comment]
I don't understand how it can't be obvious to every maintainer they can close PR's, it's literally one of their buttons

**Comment by ninjanomnom** [comment]
@MrStonedOne Yeah that's basically the intent I was going for but it could stand to be more clear. Your version I think though isn't clear enough that the pr will likely be closed as a result. I'll take a look at phrasing it better when I wake up tomorrow or someone else could try.

**Comment by ninjanomnom** [comment]
@lordpidey draft prs are just unmergable wip prs. So no difference

**Comment by fludd12** [comment]
I agree with the guideline, though I must ask if my current style of releasing crossbreeding PRs as [WIP] with effective, testable code violates them. I understand that `If your pull request is not finished make sure it is at least testable in a live environment` doesn't apply, but I would not mind finishing the vast majority of a crossbreed change before PR'ing in the future, if it is preferable. My current standard is keep additional extracts in their own commit.

**Comment by ninjanomnom** [comment]
If the pr is testable then it's fine. Having a testable pr adding a new system then using the wip to convert/add in new things using that system is a fine pattern I see no problem with.

File: .github/CONTRIBUTING.md [hunk 1]
```diff
@@ -394,6 +394,10 @@ There is no strict process when it comes to merging pull requests. Pull requests
 
 * Please explain why you are submitting the pull request, and how you think your change will be beneficial to the game. Failure to do so will be grounds for rejecting the PR.
 
+* If your pull request is not finished make sure it is at least testable in a live environment. Pull requests that do not at least meet this requirement will be closed. You may request a maintainer reopen the pull request when you're ready, or make a new one.
+
+* Maintainers may close pull requests that are deemed to be substantialy flawed. For example pull requests which would need major rewrites to match our quality guidelines. You should take some time to discuss with maintainers or other contributors on how to improve the change.
```
[the remark this answers is not in the document]
Comment (Whoneedspacee) [reply .github/CONTRIBUTING.md]: FIX IT SHOULD BE SUBSTANTIALLY

[the remark this answers is not in the document]
Comment (ExcessiveUseOfCobblestone) [reply .github/CONTRIBUTING.md]: we need to demand people use the suggestion feature for spelling corrections, what a sin!


```suggestion
* Maintainers may close pull requests that are deemed to be substantially flawed. For example pull requests which would need major rewrites to match our quality guidelines. You should take some time to discuss with maintainers or other contributors on how to improve the change.
```

[the remark this answers is not in the document]
Comment (Whoneedspacee) [reply .github/CONTRIBUTING.md]: mobile

[the remark this answers is not in the document]
Comment (ExcessiveUseOfCobblestone) [reply .github/CONTRIBUTING.md]: even worse

**Comment by Qustinnus** [comment]
the main issue is that creating discussion about branches is almost impossible if I don't have a PR open; If I try to discuss ideas that aren't in a PR I will not get any feedback or information, trying to ping maintainers or specific coders never really achieves anything because usually I get a "whatever"-esque reply or no reply at all.

besides that; I use PRs as a funnel for feedback because the subforums for coding are not used by everyone and PRs make an announcement in the server. Most of the time my PRs are shaped on feedback from people commenting on the PR; and if I only PR at the end I will most likely have to replace large parts of my work.

**Comment by ninjanomnom** [comment]
You can discuss with others, not necessarily maintainers, about the overall concept if it's so incomplete that you're still at the design stage. Figure out the broad strokes of how you're going to accomplish all you want before hand then just write the code as best you can so it's ready for testmerging.

This isn't banning all WIP prs, this is banning WIP prs that are so unfinished they're pointless to look at and just waste time.

**Pull request merged**